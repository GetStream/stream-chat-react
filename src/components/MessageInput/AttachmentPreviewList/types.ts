import { LocalAttachment } from '../types';
import type { DefaultStreamChatGenerics } from '../../../types';

export type AttachmentPreviewProps<
  A extends LocalAttachment,
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  attachment: A;
  handleRetry: (
    attachment: LocalAttachment<StreamChatGenerics>,
  ) => void | Promise<LocalAttachment<StreamChatGenerics> | undefined>;
  removeAttachments: (ids: string[]) => void;
};
