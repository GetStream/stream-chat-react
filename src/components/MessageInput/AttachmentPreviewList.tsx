import React, { useCallback } from 'react';
import { FileIcon } from 'react-file-utils';

import { useMessageInputContext } from '../../context/MessageInputContext';
import { useFileState } from './hooks/useFileState';

import { CloseIcon, DownloadIcon, LoadingIndicatorIcon, RetryIcon } from './icons';

export const AttachmentPreviewList = () => {
  const { fileOrder, imageOrder } = useMessageInputContext('AttachmentPreviewList');

  return (
    <div className='str-chat__attachment-preview-list'>
      <div
        className='str-chat__attachment-list-scroll-container'
        data-testid='attachment-list-scroll-container'
      >
        {imageOrder.map((id) => (
          <ImagePreviewItem id={id} key={id} />
        ))}
        {fileOrder.map((id) => (
          <FilePreviewItem id={id} key={id} />
        ))}
      </div>
    </div>
  );
};

type PreviewItemProps = { id: string };

const ImagePreviewItem = ({ id }: PreviewItemProps) => {
  const { imageUploads, removeImage, uploadImage } = useMessageInputContext('ImagePreviewItem');

  const handleRemove: React.MouseEventHandler<HTMLButtonElement> = useCallback(
    (e) => {
      e.stopPropagation();
      removeImage(id);
    },
    [removeImage, id],
  );
  const handleRetry = useCallback(() => uploadImage(id), [uploadImage, id]);

  const image = imageUploads[id];
  const state = useFileState(image);

  // do not display scraped attachments
  if (!image || image.og_scrape_url) return null;

  return (
    <div className='str-chat__attachment-preview-image' data-testid='attachment-preview-image'>
      <button
        className='str-chat__attachment-preview-delete'
        data-testid='image-preview-item-delete-button'
        disabled={state.uploading}
        onClick={handleRemove}
      >
        <CloseIcon />
      </button>

      {state.failed && (
        <button
          className='str-chat__attachment-preview-error str-chat__attachment-preview-error-image'
          data-testid='image-preview-item-retry-button'
          onClick={handleRetry}
        >
          <RetryIcon />
        </button>
      )}

      {state.uploading && (
        <div className='str-chat__attachment-preview-image-loading'>
          <LoadingIndicatorIcon size={17} />
        </div>
      )}

      {(image.previewUri || image.url) && (
        <img
          alt={image.file.name}
          className='str-chat__attachment-preview-thumbnail'
          src={image.previewUri ?? image.url}
        />
      )}
    </div>
  );
};

const FilePreviewItem = ({ id }: PreviewItemProps) => {
  const { fileUploads, removeFile, uploadFile } = useMessageInputContext('FilePreviewItem');

  const handleRemove: React.MouseEventHandler<HTMLButtonElement> = useCallback(
    (e) => {
      e.stopPropagation();
      removeFile(id);
    },
    [removeFile, id],
  );
  const handleRetry = useCallback(() => uploadFile(id), [uploadFile, id]);

  const file = fileUploads[id];
  const state = useFileState(file);

  if (!file) return null;

  return (
    <div className='str-chat__attachment-preview-file' data-testid='attachment-preview-file'>
      <div className='str-chat__attachment-preview-file-icon'>
        <FileIcon filename={file.file.name} mimeType={file.file.type} version='2' />
      </div>

      <button
        className='str-chat__attachment-preview-delete'
        data-testid='file-preview-item-delete-button'
        disabled={state.uploading}
        onClick={handleRemove}
      >
        <CloseIcon />
      </button>

      {state.failed && (
        <button
          className='str-chat__attachment-preview-error str-chat__attachment-preview-error-file'
          data-testid='file-preview-item-retry-button'
          onClick={handleRetry}
        >
          <RetryIcon />
        </button>
      )}

      <div className='str-chat__attachment-preview-file-end'>
        <div className='str-chat__attachment-preview-file-name'>{file.file.name}</div>
        {state.finished && (
          <a
            className='str-chat__attachment-preview-file-download'
            download
            href={file.url}
            rel='noreferrer'
            target='_blank'
          >
            <DownloadIcon />
          </a>
        )}
        {state.uploading && <LoadingIndicatorIcon size={17} />}
      </div>
    </div>
  );
};
