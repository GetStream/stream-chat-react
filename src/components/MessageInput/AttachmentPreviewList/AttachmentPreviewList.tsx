import React, { ComponentType } from 'react';
import {
  UnsupportedAttachmentPreview as DefaultUnknownAttachmentPreview,
  UnsupportedAttachmentPreviewProps,
} from './UnsupportedAttachmentPreview';
import {
  VoiceRecordingPreview as DefaultVoiceRecordingPreview,
  VoiceRecordingPreviewProps,
} from './VoiceRecordingPreview';
import {
  FileAttachmentPreview as DefaultFilePreview,
  FileAttachmentPreviewProps,
} from './FileAttachmentPreview';
import { FileUploadPreviewAdapter, ImageUploadPreviewAdapter } from './UploadPreviewItem';
import {
  ImageAttachmentPreview as DefaultImagePreview,
  ImageAttachmentPreviewProps,
} from './ImageAttachmentPreview';
import {
  isLocalAttachment,
  isLocalAudioAttachment,
  isLocalFileAttachment,
  isLocalImageAttachment,
  isLocalMediaAttachment,
  isLocalVoiceRecordingAttachment,
  isScrapedContent,
} from '../../Attachment';
import { useMessageInputContext } from '../../../context';

import type { DefaultStreamChatGenerics } from '../../../types';

export type AttachmentPreviewListProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  AudioAttachmentPreview?: ComponentType<FileAttachmentPreviewProps>;
  FileAttachmentPreview?: ComponentType<FileAttachmentPreviewProps>;
  ImageAttachmentPreview?: ComponentType<ImageAttachmentPreviewProps<StreamChatGenerics>>;
  UnsupportedAttachmentPreview?: ComponentType<
    UnsupportedAttachmentPreviewProps<StreamChatGenerics>
  >;
  VideoAttachmentPreview?: ComponentType<FileAttachmentPreviewProps>;
  VoiceRecordingPreview?: ComponentType<VoiceRecordingPreviewProps<StreamChatGenerics>>;
};

export const AttachmentPreviewList = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>({
  AudioAttachmentPreview = DefaultFilePreview,
  FileAttachmentPreview = DefaultFilePreview,
  ImageAttachmentPreview = DefaultImagePreview,
  UnsupportedAttachmentPreview = DefaultUnknownAttachmentPreview,
  VideoAttachmentPreview = DefaultFilePreview,
  VoiceRecordingPreview = DefaultVoiceRecordingPreview,
}: AttachmentPreviewListProps<StreamChatGenerics>) => {
  const {
    attachments,
    fileOrder,
    imageOrder,
    removeAttachments,
    uploadAttachment,
  } = useMessageInputContext<StreamChatGenerics>('AttachmentPreviewList');

  return (
    <div className='str-chat__attachment-preview-list'>
      <div
        className='str-chat__attachment-list-scroll-container'
        data-testid='attachment-list-scroll-container'
      >
        {attachments.map((attachment) => {
          if (isScrapedContent(attachment)) return null;
          if (isLocalVoiceRecordingAttachment(attachment)) {
            return (
              <VoiceRecordingPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id || attachment.asset_url}
                removeAttachments={removeAttachments}
              />
            );
          } else if (isLocalAudioAttachment(attachment)) {
            return (
              <AudioAttachmentPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id || attachment.asset_url}
                removeAttachments={removeAttachments}
              />
            );
          } else if (isLocalMediaAttachment(attachment)) {
            return (
              <VideoAttachmentPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id || attachment.asset_url}
                removeAttachments={removeAttachments}
              />
            );
          } else if (isLocalImageAttachment(attachment)) {
            return (
              <ImageAttachmentPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id || attachment.image_url}
                removeAttachments={removeAttachments}
              />
            );
          } else if (isLocalFileAttachment(attachment)) {
            return (
              <FileAttachmentPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id || attachment.asset_url}
                removeAttachments={removeAttachments}
              />
            );
          } else if (isLocalAttachment(attachment)) {
            return (
              <UnsupportedAttachmentPreview
                attachment={attachment}
                handleRetry={uploadAttachment}
                key={attachment.localMetadata.id}
                removeAttachments={removeAttachments}
              />
            );
          }
          return null;
        })}
        {imageOrder.map((id) => (
          <ImageUploadPreviewAdapter id={id} key={id} Preview={ImageAttachmentPreview} />
        ))}
        {fileOrder.map((id) => (
          <FileUploadPreviewAdapter id={id} key={id} Preview={FileAttachmentPreview} />
        ))}
      </div>
    </div>
  );
};
