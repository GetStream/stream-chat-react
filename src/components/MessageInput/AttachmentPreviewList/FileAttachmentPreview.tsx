import React from 'react';
import { FileIcon } from '../../ReactFileUtilities';
import { CloseIcon, DownloadIcon, LoadingIndicatorIcon, RetryIcon } from '../icons';
import { useTranslationContext } from '../../../context';

import type { AttachmentPreviewProps } from './types';
import { LocalAttachmentCast, LocalAttachmentUploadMetadata } from '../types';
import type { DefaultStreamChatGenerics } from '../../../types';

type FileLikeAttachment = {
  asset_url?: string;
  file_size?: number;
  mime_type?: string;
  title?: string;
};

export type FileAttachmentPreviewProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
  CustomLocalMetadata = Record<string, unknown>
> = AttachmentPreviewProps<
  LocalAttachmentCast<FileLikeAttachment, LocalAttachmentUploadMetadata & CustomLocalMetadata>,
  StreamChatGenerics
>;

export const FileAttachmentPreview = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>({
  attachment,
  handleRetry,
  removeAttachments,
}: FileAttachmentPreviewProps<StreamChatGenerics>) => {
  const { t } = useTranslationContext('FilePreview');
  return (
    <div className='str-chat__attachment-preview-file' data-testid='attachment-preview-file'>
      <div className='str-chat__attachment-preview-file-icon'>
        <FileIcon filename={attachment.title} mimeType={attachment.mime_type} />
      </div>

      <button
        aria-label={t('aria/Remove attachment')}
        className='str-chat__attachment-preview-delete'
        data-testid='file-preview-item-delete-button'
        disabled={attachment.localMetadata?.uploadState === 'uploading'}
        onClick={() =>
          attachment.localMetadata?.id && removeAttachments([attachment.localMetadata?.id])
        }
      >
        <CloseIcon />
      </button>

      {attachment.localMetadata?.uploadState === 'failed' && !!handleRetry && (
        <button
          className='str-chat__attachment-preview-error str-chat__attachment-preview-error-file'
          data-testid='file-preview-item-retry-button'
          onClick={() => handleRetry(attachment)}
        >
          <RetryIcon />
        </button>
      )}

      <div className='str-chat__attachment-preview-file-end'>
        <div className='str-chat__attachment-preview-file-name' title={attachment.title}>
          {attachment.title}
        </div>
        {attachment.localMetadata?.uploadState === 'finished' && !!attachment.asset_url && (
          <a
            aria-label={t('aria/Download attachment')}
            className='str-chat__attachment-preview-file-download'
            download
            href={attachment.asset_url}
            rel='noreferrer'
            target='_blank'
            title={t('Download attachment {{ name }}', { name: attachment.title })}
          >
            <DownloadIcon />
          </a>
        )}
        {attachment.localMetadata?.uploadState === 'uploading' && (
          <LoadingIndicatorIcon size={17} />
        )}
      </div>
    </div>
  );
};
