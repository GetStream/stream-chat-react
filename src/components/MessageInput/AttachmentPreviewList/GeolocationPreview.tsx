import type { LiveLocationPreview, StaticLocationPreview } from 'stream-chat';
import { CloseIcon } from '../icons';
import type { ComponentType } from 'react';
import React from 'react';
import { useTranslationContext } from '../../../context';
import { GeolocationIcon } from '../../Attachment/icons';

type GeolocationPreviewImageProps = {
  location: StaticLocationPreview | LiveLocationPreview;
};

const GeolocationPreviewImage = () => (
  <div className='str-chat__location-preview-image'>
    <GeolocationIcon />
  </div>
);

export type GeolocationPreviewProps = {
  location: StaticLocationPreview | LiveLocationPreview;
  PreviewImage?: ComponentType<GeolocationPreviewImageProps>;
  remove?: () => void;
};

export const GeolocationPreview = ({
  location,
  PreviewImage = GeolocationPreviewImage,
  remove,
}: GeolocationPreviewProps) => {
  const { t } = useTranslationContext();
  return (
    <div className='str-chat__location-preview' data-testid='location-preview'>
      <PreviewImage location={location} />
      {remove && (
        <button
          aria-label={t('aria/Remove location attachment')}
          className='str-chat__attachment-preview-delete'
          data-testid='location-preview-item-delete-button'
          onClick={remove}
        >
          <CloseIcon />
        </button>
      )}

      <div className='str-chat__attachment-preview-metadata'>
        {(location as LiveLocationPreview).durationMs ? (
          <>
            <div
              className='str-chat__attachment-preview-title'
              title={t('Shared live location')}
            >
              {t('Live location')}
            </div>
            <div className='str-chat__attachment-preview-subtitle'>
              {t('Live for {{duration}}', {
                duration: t('duration/Share Location', {
                  milliseconds: (location as LiveLocationPreview).durationMs,
                }),
              })}
            </div>
          </>
        ) : (
          <>
            <div
              className='str-chat__attachment-preview-title'
              title={t('Current location')}
            >
              {t('Current location')}
            </div>
            <div className='str-chat__attachment-preview-subtitle'>
              {location.latitude}, {location.longitude}
            </div>
          </>
        )}
      </div>
    </div>
  );
};
