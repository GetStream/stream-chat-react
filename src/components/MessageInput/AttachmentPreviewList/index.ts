export * from './AttachmentPreviewList';
export type { FileAttachmentPreviewProps } from './FileAttachmentPreview';
export type { ImageAttachmentPreviewProps } from './ImageAttachmentPreview';
export type { AttachmentPreviewProps } from './types';
export type { UnsupportedAttachmentPreviewProps } from './UnsupportedAttachmentPreview';
export type { VoiceRecordingPreviewProps } from './VoiceRecordingPreview';
