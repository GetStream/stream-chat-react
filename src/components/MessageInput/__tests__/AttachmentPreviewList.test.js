/* eslint-disable jest-dom/prefer-in-document */

import React from 'react';

import { fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';

import { AttachmentPreviewList } from '../AttachmentPreviewList';
import { MessageInputContextProvider } from '../../../context/MessageInputContext';

import { generateUpload } from '../../../mock-builders';

const uploadsReducer = (pv, cv) => {
  pv[cv.id] = cv;
  return pv;
};

/**
 * @param {string} word
 * @returns string
 */
const capitalize = ([firstLetter, ...restOfTheWord]) =>
  `${firstLetter.toUpperCase()}${restOfTheWord.join('')}`;

const orderMapper = ({ id }) => id;

const generateMessageInputContextValue = ({ files = [], images = [] } = {}) => ({
  fileOrder: files.map(orderMapper),
  fileUploads: files.reduce(uploadsReducer, {}),
  imageOrder: images.map(orderMapper),
  imageUploads: images.reduce(uploadsReducer, {}),
  removeFile: jest.fn(),
  removeImage: jest.fn(),
  uploadFile: jest.fn(),
  uploadImage: jest.fn(),
});

const renderComponent = (value = {}, renderFunction = render) =>
  renderFunction(
    <MessageInputContextProvider value={{ ...generateMessageInputContextValue(), ...value }}>
      <AttachmentPreviewList />
    </MessageInputContextProvider>,
  );

jest.mock('nanoid', () => ({
  nanoid: () => '<randomNanoId>',
}));

describe('AttachmentPreviewList', () => {
  it('renders without any attachments', () => {
    const { getByTestId } = renderComponent();

    const attachmentList = getByTestId('attachment-list-scroll-container');

    expect(attachmentList).toBeEmptyDOMElement();
  });

  it.each(['uploading', 'failed', 'finished'])(
    'renders with one image and one file with state "%s"',
    (state) => {
      const [file, image] = [
        generateUpload({
          fileOverrides: { name: 'file-upload' },
          objectOverrides: { state },
        }),
        generateUpload({
          fileOverrides: { name: 'image-upload', type: 'image' },
          objectOverrides: { state },
        }),
      ];

      const tree = renderComponent(
        generateMessageInputContextValue({ files: [file], images: [image] }),
        renderer.create,
      ).toJSON();

      expect(tree).toMatchSnapshot();
    },
  );

  it.each(['file', 'image'])('tests "retry" click on %s upload', (type) => {
    const file = generateUpload({
      fileOverrides: { type },
      objectOverrides: { state: 'failed' },
    });

    const contextValue = generateMessageInputContextValue({ [`${type}s`]: [file] });

    const { getByTestId } = renderComponent(contextValue);

    const retryButton = getByTestId(`${type}-preview-item-retry-button`);

    fireEvent.click(retryButton);

    expect(contextValue[`upload${capitalize(type)}`]).toHaveBeenCalledWith(file.id);
  });

  it.each(['file', 'image'])('tests "remove" click on %s upload', (type) => {
    const file = generateUpload({
      fileOverrides: { type },
      objectOverrides: { state: 'finished' },
    });

    const contextValue = generateMessageInputContextValue({ [`${type}s`]: [file] });

    const { getByTestId } = renderComponent(contextValue);

    const deleteButton = getByTestId(`${type}-preview-item-delete-button`);

    fireEvent.click(deleteButton);

    expect(contextValue[`remove${capitalize(type)}`]).toHaveBeenCalledWith(file.id);
  });
});
