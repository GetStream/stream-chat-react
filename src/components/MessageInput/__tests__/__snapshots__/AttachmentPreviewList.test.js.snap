// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentPreviewList renders with one image and one file with state "failed" 1`] = `
<div
  className="str-chat__attachment-preview-list"
>
  <div
    className="str-chat__attachment-list-scroll-container"
    data-testid="attachment-list-scroll-container"
  >
    <div
      className="str-chat__attachment-preview-image"
      data-testid="attachment-preview-image"
    >
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="image-preview-item-delete-button"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <button
        className="str-chat__attachment-preview-error str-chat__attachment-preview-error-image"
        data-testid="image-preview-item-retry-button"
        onClick={[Function]}
      >
        <svg
          data-testid="retry"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.6449 6.35C16.1949 4.9 14.2049 4 11.9949 4C7.57488 4 4.00488 7.58 4.00488 12C4.00488 16.42 7.57488 20 11.9949 20C15.7249 20 18.8349 17.45 19.7249 14H17.6449C16.8249 16.33 14.6049 18 11.9949 18C8.68488 18 5.99488 15.31 5.99488 12C5.99488 8.69 8.68488 6 11.9949 6C13.6549 6 15.1349 6.69 16.2149 7.78L12.9949 11H19.9949V4L17.6449 6.35Z"
            fill="black"
          />
        </svg>
      </button>
      <img
        alt="image-upload"
        className="str-chat__attachment-preview-thumbnail"
        src="url"
      />
    </div>
    <div
      className="str-chat__attachment-preview-file"
      data-testid="attachment-preview-file"
    >
      <div
        className="str-chat__attachment-preview-file-icon"
      >
        <svg
          className="rfu-file-fallback"
          filename="file-upload"
          fill="none"
          height={20}
          viewBox="0 0 34 40"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 3C0 1.34315 1.34315 0 3 0H23L34 11V37C34 38.6569 32.6569 40 31 40H3C1.34315 40 0 38.6569 0 37V3Z"
            fill="url(#paint0_linear)"
          />
          <path
            d="M34 11L26 11C24.3431 11 23 9.65685 23 8V0L34 11Z"
            fill="#DBDBDB"
          />
          <path
            clipRule="evenodd"
            d="M8 13H23V15H8V13Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 17H18V19H8V17Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 21H23V23H8V21Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <defs>
            <linearGradient
              gradientUnits="userSpaceOnUse"
              id="paint0_linear"
              x1="0"
              x2="0"
              y1="0"
              y2="40"
            >
              <stop
                stopColor="white"
              />
              <stop
                offset="1"
                stopColor="#DBDBDB"
              />
            </linearGradient>
          </defs>
        </svg>
      </div>
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="file-preview-item-delete-button"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <button
        className="str-chat__attachment-preview-error str-chat__attachment-preview-error-file"
        data-testid="file-preview-item-retry-button"
        onClick={[Function]}
      >
        <svg
          data-testid="retry"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.6449 6.35C16.1949 4.9 14.2049 4 11.9949 4C7.57488 4 4.00488 7.58 4.00488 12C4.00488 16.42 7.57488 20 11.9949 20C15.7249 20 18.8349 17.45 19.7249 14H17.6449C16.8249 16.33 14.6049 18 11.9949 18C8.68488 18 5.99488 15.31 5.99488 12C5.99488 8.69 8.68488 6 11.9949 6C13.6549 6 15.1349 6.69 16.2149 7.78L12.9949 11H19.9949V4L17.6449 6.35Z"
            fill="black"
          />
        </svg>
      </button>
      <div
        className="str-chat__attachment-preview-file-end"
      >
        <div
          className="str-chat__attachment-preview-file-name"
        >
          file-upload
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AttachmentPreviewList renders with one image and one file with state "finished" 1`] = `
<div
  className="str-chat__attachment-preview-list"
>
  <div
    className="str-chat__attachment-list-scroll-container"
    data-testid="attachment-list-scroll-container"
  >
    <div
      className="str-chat__attachment-preview-image"
      data-testid="attachment-preview-image"
    >
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="image-preview-item-delete-button"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <img
        alt="image-upload"
        className="str-chat__attachment-preview-thumbnail"
        src="url"
      />
    </div>
    <div
      className="str-chat__attachment-preview-file"
      data-testid="attachment-preview-file"
    >
      <div
        className="str-chat__attachment-preview-file-icon"
      >
        <svg
          className="rfu-file-fallback"
          filename="file-upload"
          fill="none"
          height={20}
          viewBox="0 0 34 40"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 3C0 1.34315 1.34315 0 3 0H23L34 11V37C34 38.6569 32.6569 40 31 40H3C1.34315 40 0 38.6569 0 37V3Z"
            fill="url(#paint0_linear)"
          />
          <path
            d="M34 11L26 11C24.3431 11 23 9.65685 23 8V0L34 11Z"
            fill="#DBDBDB"
          />
          <path
            clipRule="evenodd"
            d="M8 13H23V15H8V13Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 17H18V19H8V17Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 21H23V23H8V21Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <defs>
            <linearGradient
              gradientUnits="userSpaceOnUse"
              id="paint0_linear"
              x1="0"
              x2="0"
              y1="0"
              y2="40"
            >
              <stop
                stopColor="white"
              />
              <stop
                offset="1"
                stopColor="#DBDBDB"
              />
            </linearGradient>
          </defs>
        </svg>
      </div>
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="file-preview-item-delete-button"
        disabled={false}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <div
        className="str-chat__attachment-preview-file-end"
      >
        <div
          className="str-chat__attachment-preview-file-name"
        >
          file-upload
        </div>
        <a
          className="str-chat__attachment-preview-file-download"
          download={true}
          href="url"
          rel="noreferrer"
          target="_blank"
        >
          <svg
            data-testid="download"
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19.35 10.04C18.67 6.59 15.64 4 12 4C9.11 4 6.6 5.64 5.35 8.04C2.34 8.36 0 10.91 0 14C0 17.31 2.69 20 6 20H19C21.76 20 24 17.76 24 15C24 12.36 21.95 10.22 19.35 10.04ZM19 18H6C3.79 18 2 16.21 2 14C2 11.95 3.53 10.24 5.56 10.03L6.63 9.92L7.13 8.97C8.08 7.14 9.94 6 12 6C14.62 6 16.88 7.86 17.39 10.43L17.69 11.93L19.22 12.04C20.78 12.14 22 13.45 22 15C22 16.65 20.65 18 19 18ZM13.45 10H10.55V13H8L12 17L16 13H13.45V10Z"
              fill="black"
            />
          </svg>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`AttachmentPreviewList renders with one image and one file with state "uploading" 1`] = `
<div
  className="str-chat__attachment-preview-list"
>
  <div
    className="str-chat__attachment-list-scroll-container"
    data-testid="attachment-list-scroll-container"
  >
    <div
      className="str-chat__attachment-preview-image"
      data-testid="attachment-preview-image"
    >
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="image-preview-item-delete-button"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <div
        className="str-chat__attachment-preview-image-loading"
      >
        <div
          className="str-chat__loading-indicator"
        >
          <svg
            data-testid="loading-indicator"
            height={17}
            viewBox="0 0 30 30"
            width={17}
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <linearGradient
                id="<randomNanoId>-linear-gradient"
                x1="50%"
                x2="50%"
                y1="0%"
                y2="100%"
              >
                <stop
                  offset="0%"
                  stopColor="#FFF"
                  stopOpacity="0"
                />
                <stop
                  data-testid="stop-color"
                  offset="100%"
                  stopOpacity="1"
                />
              </linearGradient>
            </defs>
            <path
              d="M2.518 23.321l1.664-1.11A12.988 12.988 0 0 0 15 28c7.18 0 13-5.82 13-13S22.18 2 15 2V0c8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15-5.206 0-9.792-2.652-12.482-6.679z"
              fill="url(#<randomNanoId>-linear-gradient)"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
      <img
        alt="image-upload"
        className="str-chat__attachment-preview-thumbnail"
        src="url"
      />
    </div>
    <div
      className="str-chat__attachment-preview-file"
      data-testid="attachment-preview-file"
    >
      <div
        className="str-chat__attachment-preview-file-icon"
      >
        <svg
          className="rfu-file-fallback"
          filename="file-upload"
          fill="none"
          height={20}
          viewBox="0 0 34 40"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 3C0 1.34315 1.34315 0 3 0H23L34 11V37C34 38.6569 32.6569 40 31 40H3C1.34315 40 0 38.6569 0 37V3Z"
            fill="url(#paint0_linear)"
          />
          <path
            d="M34 11L26 11C24.3431 11 23 9.65685 23 8V0L34 11Z"
            fill="#DBDBDB"
          />
          <path
            clipRule="evenodd"
            d="M8 13H23V15H8V13Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 17H18V19H8V17Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <path
            clipRule="evenodd"
            d="M8 21H23V23H8V21Z"
            fill="#CFCFCF"
            fillRule="evenodd"
          />
          <defs>
            <linearGradient
              gradientUnits="userSpaceOnUse"
              id="paint0_linear"
              x1="0"
              x2="0"
              y1="0"
              y2="40"
            >
              <stop
                stopColor="white"
              />
              <stop
                offset="1"
                stopColor="#DBDBDB"
              />
            </linearGradient>
          </defs>
        </svg>
      </div>
      <button
        className="str-chat__attachment-preview-delete"
        data-testid="file-preview-item-delete-button"
        disabled={true}
        onClick={[Function]}
      >
        <svg
          data-testid="close-no-outline"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12L19 6.41Z"
            fill="black"
          />
        </svg>
      </button>
      <div
        className="str-chat__attachment-preview-file-end"
      >
        <div
          className="str-chat__attachment-preview-file-name"
        >
          file-upload
        </div>
        <div
          className="str-chat__loading-indicator"
        >
          <svg
            data-testid="loading-indicator"
            height={17}
            viewBox="0 0 30 30"
            width={17}
            xmlns="http://www.w3.org/2000/svg"
          >
            <defs>
              <linearGradient
                id="<randomNanoId>-linear-gradient"
                x1="50%"
                x2="50%"
                y1="0%"
                y2="100%"
              >
                <stop
                  offset="0%"
                  stopColor="#FFF"
                  stopOpacity="0"
                />
                <stop
                  data-testid="stop-color"
                  offset="100%"
                  stopOpacity="1"
                />
              </linearGradient>
            </defs>
            <path
              d="M2.518 23.321l1.664-1.11A12.988 12.988 0 0 0 15 28c7.18 0 13-5.82 13-13S22.18 2 15 2V0c8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15-5.206 0-9.792-2.652-12.482-6.679z"
              fill="url(#<randomNanoId>-linear-gradient)"
              fillRule="evenodd"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
